E2E libraryzation

jayunit100 opened the pull request.

jayunit100 committed:
E2E libraryzation with removal of "init()" flags from _test.go - and Setup configuration error handling for extenders.

jayunit100 commented:
For #19461

liggitt commented on test/e2e/e2e_test_lib.go:
```diff
@@ -0,0 +1,188 @@
+	reportDir = flag.String("report-dir", "", "Path to the directory where the JUnit XML reports should be saved. Default is empty, which doesn't generate these reports.")
+)
+
+func init() {
```
can we put these in a non-init function if this will be part of non-test code?

liggitt commented:
I would just call the file `e2e.go`

jayunit100 commented:
@liggitt agree on the `init()`, we can implement it at runtime rather than using the `init()` machinery, fixing now.

jayunit100 commented:
Agree re: flags.  I've cleaned that init stuff up so that its controlled by the calling _test function.  The additional advantage here is that other libs can change the flags if they need to before parsing them.

jayunit100 commented:
Okay, Ive Rebased and tested.  Also `init()` refactored  (retained as second commit).  
PTAL :)

timothysc commented:
/cc @kubernetes/goog-testing

timothysc commented:
The purpose of this pr is to make it easier for downstream extensibility of the e2e(s) for both openshift and potentially the contrib and possibly other supporting repositories.  

This would be a 1st step. 

/cc @karlkfi @quinton-hoole

timothysc commented:
@jayunit100 please squash commits.

jayunit100 commented:
Ok.  Updated and squashed the patch.  
- Moves the old e2e_test into e2e.go
- Adds the thin wrapper which can be easily customized by openshift or other extenders.
- Adds a fast failure if kubeconfig is missing, with extensible method for checking setup before tests start.
- Adds all ginkgo configuration into the calling function, rather than the e2e.go file.  This makes it again easier for extenders.

Ready for a final review.

pmorie commented on test/e2e/e2e.go:
```diff
@@ -0,0 +1,202 @@
+// RunE2ETest is a test function which can be called externally.  This facilitates usage of the Kubernetes E2E tests in different environments.
+func (e2e *E2ETests) RunE2ETest(t *testing.T) {
+	if e2e.IsSetup() == false {
+		glog.Fatal("Essential inputs for running e2e tests was not provided.  See warnings above for details.")
```
s/was/were/

pmorie commented on test/e2e/e2e.go:
```diff
@@ -0,0 +1,202 @@
+		glog.Info("The --provider flag is not set.  Treating as a conformance test.  Some tests may not be run.")
+	}
+
+	if testContext.Provider == "gce" || testContext.Provider == "gke" {
```
does GKE use the gcecloud package?

pmorie replied:
Nevermind, I see it in the diff now.

pmorie commented on test/e2e/e2e.go:
```diff
@@ -0,0 +1,202 @@
+)
+
+var (
+	cloudConfig = &testContext.CloudConfig
```
Should this be exported?

jayunit100 replied:
Copied from previous impl.

pmorie replied:
I think the downstream user needs to be able to alter this config.  I'd prefer not to share state via a variable, though -- inject it as an arg to `RunE2ETests`?

jayunit100 replied:
Yup, I agree.  Good point.  Im going to make this configurable.

jayunit100 replied:
This is a bit of a rathole after digging.  I've avoided making any changes in this PR and made a follow on issue to further decouple things. For now, SetTestContext is available for last minute rejiggering of the global test props

pmorie commented on test/e2e/e2e.go:
```diff
@@ -0,0 +1,202 @@
+	// podStartupTimeout is the time to allow all pods in the cluster to become
+	// running and ready before any e2e tests run. It includes pulling all of
+	// the pods (as of 5/18/15 this is 8 pods).
+	podStartupTimeout = 10 * time.Minute
```
Should this be exported?

jayunit100 replied:
Copied from previous impl.

pmorie commented:
Generally, this LGTM, I just have a couple Qs about whether certain symbols should be exported

ixdy commented:
haven't had a chance to look at this yet, sorry. will do so tomorrow.

jayunit100 commented:
okay.  I've updated with @pmorie 's typo fix also.

jayunit100 commented:
@pmorie i tend to agree some cleanup on CloudConfig/ReportDir could be done .  But for now, the cmd line options make it debatable wether or not its needed for this patch.

How do you feel about a separate issue for these structifying all the parameters in E2E?

ixdy commented on test/e2e/e2e.go:
```diff
@@ -0,0 +1,202 @@
+}
+
+// Sanity checks that the e2e's have the basic information required to run before launching them.
+func (e2e *E2ETests) IsSetup() bool {
```
does this need to be factored out of `RunE2ETests`? And if so, does it need to be exported, too?

jayunit100 replied:
I don't think so. The purpose is to make sure that when we call run , the user has set the command line switches appropriately

jayunit100 replied:
there are plans to deflag the e2es eventually...  The is Setup deparates the "are flags set" from "is it ready to run"... The latter is the more important question .  Does that make sense ?

ixdy replied:
I guess I'm just confused at what you pulled into this func vs what you left in `RunE2ETest`. why check KubeConfig here but everything else there?

jayunit100 replied:
Totally agree. KubeConfig is a sanity check that is super important.  I didn't want to do a massive diff for this patch, but yeah, IsSetup is a much better place for All the data level error checking .

jayunit100 replied:
@ixdy in the follow on Jira where we improve testContext, I'm thinking all the config checking can be externalized. Again the goal here is to minimally modify the e2e so that if they are misconfigured by an extending component, there is a fast failure message .   

Context: The likelihood of such misconfiguration happening is a little higher now, since the e2e'go no longer uses "init", the ordering of config is not enforced for us by golang, but rather, by the e2e_test.go file...

...So I've just externalized the validation so that it preempts noisy failures that happen when flag bindings weren't  called properly in order .

ixdy replied:
in the issue you linked [you gave an example](https://github.com/kubernetes/kubernetes/issues/19461#issuecomment-170681821) in which only `RunE2ETest` is called, which suggests that that is the only external API - so I still don't understand why `IsSetup` and `RegisterE2EFlags` need to be exported.

ixdy commented on test/e2e/e2e.go:
```diff
@@ -0,0 +1,202 @@
+	return true
+}
+
+func (e2e *E2ETests) RegisterE2EFlags() {
```
I'm a little confused as to why this is a function on the struct, as it doesn't do anything with it...

ixdy replied:
also, does it need to be exported?

jayunit100 replied:
I've put this on the struct because over time we want the struct  to hold more info, and this allows us to contextualize the fact that all this method does is make sure the configuration params are written to an object.  

Right now that object is external to the e2e struct,  but in general there is no reason to register these flags outside the context of an e2e struct.

Let me know if that justifies it to you. If not I can make it a global function but I think this helps to move us away from global config dependencies , and towards a purely data-coupled configuration flow.

ixdy commented on test/e2e/e2e.go:
```diff
@@ -0,0 +1,202 @@
+}
+
+// RunE2ETest is a test function which can be called externally.  This facilitates usage of the Kubernetes E2E tests in different environments.
+func (e2e *E2ETests) RunE2ETest(t *testing.T) {
```
similarly, this doesn't use the struct

jayunit100 replied:
e2e.IsSetup is used.
That is a precondition that should trip when anyone extending the e2e suite .

ixdy commented:
OK, I'm still not sure whether you need to export everything you're currently exporting. What is the interface that external callers are expected to use?

jayunit100 commented:
- I haven't defined an explicit interface.

- In this patch, the usage is demo'd in e2e_test.go : call RegisterFlags, parse the flags, and then the tests are run.

- You are right hat IsSetup can get by without being exported... But I assume callers will want some way to know wether tests are setup or not which doesnt require knowledge of the config object.

If making IsSetup private is still an issue, and that solves your concern , I can easily do that ...

ixdy commented:
Basically, this PR is doing two things:
1) Moving the test setup/runner code out of a _test.go file so that another _test.go file can call it
2) Refactor some parts of this code into separate functions

I understand the first and it makes sense, but I'm not sure I'm following the second part completely. It feels like a lot of this is a start to a framework, with the rest getting implemented later, but without seeing how that framework is going to be used, it's hard to judge whether this design is correct.

jayunit100 commented:
@ixdy I agree, i think we've gotten carried away with minor improvements, mostly based around maximizing future configuration and removing the reliance on `init()`.

I will try to recreate a minimal patch and move these extra changes to a follow on PR..

Is it okay to keep `init` for now, and minimize all changes  @liggitt ,  @pmorie ?  

This will "work" as is, and I think the back and forth on layering more changes into this first patch, which really only needs to do one simple thing, is unnecessary.

liggitt commented:
No init() registering flags in non-"_test" files

jayunit100 commented:
Closing this PR.  Will reopen after 19624 merges.
